update_repo.sh

```diff
@@ -53,23 +70,41 @@ check_repo()
 
 update_repo()
 {
-    OPTS=
-    [ -r "$1/comps.xml" ] && OPTS="-g comps.xml"
-    $createrepo $OPTS --update -x 'fc*/*' -o $1 $1 >/dev/null
-    if [ -x "$mkmetalink" -a -r "$mirrors_list" ]; then
-        $mkmetalink $mirrors_list "$1/repodata/repomd.xml" \
+    local OPTS= tosign sign_key
+
+    if [ "x${SIGN_KEY+test}" = xtest ]; then
+        sign_key=$SIGN_KEY
+    else
+        sign_key=$(rpm --eval '%{?_gpg_name:%_gpg_name}') || return
+    fi
+    if [ -n "$sign_key" ]; then
+        [ -r "$1/comps.xml" ] && OPTS="-g comps.xml"
+        "$createrepo" $OPTS --update -x 'fc*/*' -o "$1" -- "$1" >/dev/null || return
+    fi
+    if [ -x "$mkmetalink" ] && [ -r "$mirrors_list" ]; then
+        "$mkmetalink" -- "$mirrors_list" "$1/repodata/repomd.xml" \
             > "$1/repodata/repomd.xml.metalink"
     fi
+
+    if [ "x${GNUPG+test}" = xtest ]; then
+        GNUPG=@$GNUPG
```

Comment:
Why prepending with `@`? If that prevents some specific realistic attack (configuration file is not a threat - it is trusted anyway), add a comment how. Otherwise, really, making things more complex doesn't make them more secure. I already spent a couple of minutes wondering why `${GNUPG#@}` below.

You may also want to read https://www.qubes-os.org/doc/coding-style/#general-programming-style-guidelines

Reply:
The purpose of the `@` is to distinguish the case where `GNUPG` is unset from the case where it is set to the empty string.  I use `GNUPG= SIGN_KEY=` as an explicit request to disable both metadata generation and signing.  However, this is ugly and very confusing.  Adding a separate environment variable (such as `NO_SIGN`) is a much better idea.

More generally, the purpose of the added validation code is so that if someone (like me) accidentally passes nonsensical inputs (such as a release name with a space in it), the code will fail fast with a (hopefully) useful error message, rather than silently doing the wrong thing.

Reply:
>  Adding a separate environment variable (such as `NO_SIGN`) is a much better idea.

Yes, especially as we already use it in few other places.

> More generally, the purpose of the added validation code is so that if someone (like me) accidentally passes nonsensical inputs (such as a release name with a space in it), the code will fail fast with a (hopefully) useful error message, rather than silently doing the wrong thing.

Yes, the arguments validation is a bit verbose, but its purpose is clear.